### PR TITLE
feat: Awesome project, but it is not good

### LINKED ISSUE #7: Awesome project, but it is not good
I want to add more math expressions. Do you think it is possible to implement Newton's method?

### PR DESCRIPTION
**Pull request by dseredkin**
## Summary
Automated implementation for issue #7.

## Changes
- `src/calculator.py`
- `tests/test_calculator.py`

## Original Issue
I want to add more math expressions. Do you think it is possible to implement Newton's method?

---
Generated by Code Agent

Closes #7

### CODE REVIEW

**Review by ghost — CHANGES_REQUESTED** [review]
## Review Summary
The pull request correctly implements Newton's method and meets the requirements, but the failing CI tests require fixes before approval.

### Issues Found
- CI tests are failing, which indicates potential issues with the test suite or implementation that need to be resolved.
- The code implements Newton's method as requested, but the failing CI suggests there may be bugs or unhandled edge cases.
- CI checks are failing

### Suggestions
- Investigate the CI failure logs to identify and fix the specific test issues.
- Add more comprehensive tests for edge cases, such as very small tolerances or functions with multiple roots.
- Consider adding input validation in newtons_method to ensure f and df are callable, to prevent runtime errors.

### Line Comments: 1
Please resolve the inline comments below.

---
**Requirements Met:** Yes
**CI Passing:** No

*Generated by Reviewer Agent*

File: src/calculator.py [hunk 1]
```diff
@@ -1,11 +1,36 @@
...
-    """Subtract b from a."""
-    return a - b
+    Returns:
+        The approximated root of the function.
+
+    Raises:
+        ValueError: If the method does not converge or if the derivative is zero.
+    """
+    x = x0
+    for _ in range(max_iter):
```
Comment (ghost) [inline src/calculator.py:36]: Direct equality check (dfx == 0) for floating-point numbers may not be reliable due to precision errors. Consider using abs(dfx) < 1e-10 or a similar epsilon value.

**Review by ghost — CHANGES_REQUESTED** [review]
## Review Summary
The pull request successfully implements Newton's method as requested in the original issue. The code is correct, well-structured, and adequately tested. No major issues were found, though a minor suggestion for handling floating-point equality is noted.

### Suggestions
- Consider using a small epsilon value for floating-point comparisons in the derivative check to handle potential precision issues.

### Line Comments: 1
Please resolve the inline comments below.

---
**Requirements Met:** Yes
**CI Passing:** Yes

*Generated by Reviewer Agent*

File: src/calculator.py [hunk 2]
```diff
@@ -1,11 +1,36 @@
...
+        fx = f(x)
```
Comment (ghost) [inline src/calculator.py:37]: The check 'if dfx == 0' may not handle floating-point precision errors effectively. Consider using 'if abs(dfx) < 1e-10' for better robustness.

File: tests/test_calculator.py [hunk 3]
```diff
@@ -1,13 +1,27 @@
...
+    """Test Newton's method with a simple quadratic function."""
 
+    def f(x: float) -> float:
+        return x**2 - 4  # Roots at 2 and -2
 
-def test_subtract():
-    assert subtract(5, 3) == 2
-    assert subtract(0, 5) == -5
+    def df(x: float) -> float:
+        return 2 * x
```
Comment (ghost) [inline tests/test_calculator.py:14]: The tolerance of 0.01 in the assertion might be too lenient for verifying the root approximation. Consider a smaller value like 1e-5 for more precision.

**Review by ghost — CHANGES_REQUESTED** [review]
## Review Summary
The pull request successfully implements Newton's method as per the original issue, with clean code, proper documentation, and adequate testing. No critical issues were found, and CI is passing.

### Suggestions
- Consider using a small epsilon (e.g., 1e-10) to check if the derivative is near zero instead of exactly zero to handle floating-point precision issues.
- In the tests, use a more precise tolerance in assertions (e.g., 1e-5) if the method is expected to converge accurately, to make tests more robust.

### Line Comments: 2
Please resolve the inline comments below.

---
**Requirements Met:** Yes
**CI Passing:** Yes

*Generated by Reviewer Agent*
